### PR TITLE
Fix UartMux to be able to use multiple muxes

### PR DESCRIPTION
**Pull request by alexandruradovici**
### Pull Request Overview

This pull request fixes the way in which `UartMux` is being instantiated. Before this PR:
- the `new` function from the `UartMuxComponent` would instantiate the `UartMux` using a `static_init!` macro. Due to this, instantiating two `UartMux` structures would result in the second instance overwriting the first one.
- the `RX_BUF` was defined within the `virtual_uart` capsule, meaning that two instances of `UartMux` would use the same buffer.

This PR:
- moves the instantiating of the `UartMux` in the `uart_mux_component_helper` macro.
- moves the instantiating of the `RX_BUF` in the `uart_mux_component_helper` macro.

#### Example

```rust
let uart_mux1 = components::console::UartMuxComponent::new(
    &peripherals.usart1,
    115200,
    dynamic_deferred_caller,
)
.finalize(components::uart_component_helper!(64));

let uart_mux2 = components::console::UartMuxComponent::new(
    &peripherals.usart2,
    115200,
    dynamic_deferred_caller,
)
.finalize(components::uart_component_helper!(64));
```

Without this PR `uart_mux2` overwrites `uart_mux1`, as the structure is allocated in the same memory space.

This PR enables the `usart2` for the STM32F3Discovery board.

This is similar to what the I2C and SPI components do.

### Testing Strategy

This pull request was tested using an STM32F3 Discovery board with two serial ports. One serial port was used for the kernel messages and process console, while the second one was used for applications.

### TODO or Help Wanted

N/A

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by bradjc** [comment]
Does this conflict or overlap #3046?

**Comment by alexandruradovici** [comment]
> Does this conflict or overlap #3046?

I don't think it does, #3046 does not seem to change anything for the mux initialization or the uart receive buffer. The `boards/components/src/console.rs` file is not changed.

**Review by hudson-ayers — APPROVED** [review]
Looks good to me -- it was always a problem that the receive buffer was defined in the capsule.

I am adding a new label for this PR to track changes that modify the board interface in a way that an out-of-tree board maintainer would need to replicate, just so we can track these types of changes between releases.

**Review by bradjc — CHANGES_REQUESTED** [review]
I see that this is inline with the I2C component, but I'm not sure about this change. `static_ini_half!` was created for a very specific purpose: to support sharing components when the actual object is templated on a chip-specific type. Here there is no chip-specific type in the virtual uart object, and as such the console component didn't use static_init_half.

I don't think we should continue to use static_init_half for a different purpose (to allow the same component to be used multiple times). Also, there is currently no indication of which components can be used multiple times. If components are going to be maintainable and usable, I think we need something more clear about how they can be used.

**Comment by hudson-ayers** [comment]
Ideally the policy should be that every component is reusable; but that would require a high-effort overhaul of all components to get there, and we have seen with https://github.com/tock/tock/pull/1618 that there isn't anyone motivated for a whole-scale components overhaul. IMO it seems like a shame to not at least make usability improvements on an as-needed basis, or to block those improvements on a very high-effort complete overhaul.

**Comment by alexandruradovici** [comment]
@bradjc are you suggesting moving the arguments of `new` to the macro?

The `UartMux` needs to be initialized in a macro, otherwise it will be overwritten every time we call `new`.

**Comment by bradjc** [comment]
I'm not suggesting we do a full change to components, or that we cannot upgrade the console component independently of all other components. What I'm saying is that `static_init_half!()` is a hack I added to allow us to share components across microcontrollers. According to our own documentation, we aren't supposed to be using static_init_half anymore:

https://github.com/tock/tock/blob/d21666de503955ab9bae7d1ed95c58b144b007e9/kernel/src/utilities/static_init.rs#L136-L142

But what I want to avoid is adding a new hack using an old hack to get re-usable components. I assume there must be some mechanism we can add or use that makes it clear that the console component is now reusable.

**Comment by alexandruradovici** [comment]
So using `static_buf` instead would work?

**Comment by alexandruradovici** [comment]
I updated it to use the `static_buf` macro.

File: capsules/src/virtual_uart.rs [hunk 1]
```diff
@@ -52,9 +52,6 @@ use kernel::hil::uart;
 use kernel::utilities::cells::{OptionalCell, TakeCell};
 use kernel::ErrorCode;
 
-const RX_BUF_LEN: usize = 64;
```
Comment (bradjc) [inline capsules/src/virtual_uart.rs:55]: This should be left here, as the recommended rx buf len from the author of this capsule.

Comment (alexandruradovici) [reply capsules/src/virtual_uart.rs:55]: Yes, that makes sense.

File: boards/stm32f3discovery/src/main.rs [hunk 2]
```diff
@@ -251,15 +251,15 @@ unsafe fn set_pin_primary_functions(
...
     });
 
-    // channel 3
-    gpio_ports.get_pin(PinId::PA02).map(|pin| {
-        pin.set_mode(stm32f303xc::gpio::Mode::AnalogMode);
-    });
+    // // channel 3
+    // gpio_ports.get_pin(PinId::PA02).map(|pin| {
+    //     pin.set_mode(stm32f303xc::gpio::Mode::AnalogMode);
+    // });
```
Comment (alistair23) [inline boards/stm32f3discovery/src/main.rs:257]: This seems like an unrelated change

File: boards/litex/arty/src/main.rs [hunk 3]
```diff
@@ -7,6 +7,7 @@
 #![cfg_attr(not(doc), no_main)]
 
 use capsules::virtual_alarm::{MuxAlarm, VirtualMuxAlarm};
+
```
Comment (alistair23) [inline boards/litex/arty/src/main.rs:10]: Extra newline

File: boards/stm32f3discovery/src/main.rs (hunk 2 above)
Comment (alexandruradovici) [reply boards/stm32f3discovery/src/main.rs]: Those are the pins used to test the second uart, reverted.

**Review by hudson-ayers — DISMISSED** [review]

**Review by alistair23 — DISMISSED** [review]

**Comment by alexandruradovici** [comment]
I rebased it to use the new `static_buf` with `MaybeUninit`. I think this is good to go.

**Review by bradjc — DISMISSED** [review]

**Review by alistair23 — DISMISSED** [review]

**Comment by alexandruradovici** [comment]
Rebased

File: boards/components/src/console.rs [hunk 4]
```diff
@@ -28,22 +28,41 @@ use kernel::create_capability;
...
-use kernel::static_init;
 
 use capsules::console::DEFAULT_BUF_SIZE;
 
-pub struct UartMuxComponent {
+#[macro_export]
+macro_rules! uart_mux_component_static {
+    () => {{
+        use capsules::virtual_uart::MuxUart;
+        use core::mem::MaybeUninit;
```
Comment (bradjc) [inline boards/components/src/console.rs:38]:
```suggestion
```

File: boards/components/src/console.rs [hunk 5]
```diff
@@ -28,22 +28,41 @@ use kernel::create_capability;
...
+        use kernel::static_buf;
+        let UART_MUX = static_buf!(MuxUart<'static>);
+        let RX_BUF = static_buf!([u8; capsules::virtual_uart::RX_BUF_LEN]);
+        (UART_MUX, RX_BUF)
+    }};
+    ($rx_buffer_len: literal) => {{
+        use capsules::virtual_uart::MuxUart;
+        use core::mem::MaybeUninit;
```
Comment (bradjc) [inline boards/components/src/console.rs:46]:
```suggestion
```

**Review by bradjc — CHANGES_REQUESTED** [review]

**Review by bradjc — APPROVED** [review]
